src/hotspot/share/utilities/nonblockingQueue.inline.hpp

```diff
@@ -116,7 +116,10 @@ void NonblockingQueue<T, next_ptr>::append(T& first, T& last) {
     // Successfully extended the queue list from old_tail to first.  No
     // other push/append could have competed with us, because we claimed
     // old_tail for extension.  We won any races with try_pop by changing
-    // away from end-marker.  So we're done.
+    // away from end-marker.  So we're done.  Note that ABA is possible;
+    // try_pop could take old_tail before our update, it gets recycled and
+    // re-added to the end, and then we successfully cmpxchg, rendering the
+    // list in _tail circular.
```

Comment:
Doesn't this contradict the "We won any races with try_pop ... so we're done"!

Reply:
The client of this class is expected to prevent ABA from occurring.  Some of the mechanisms that might be used for doing so include separate phases for push/pop and preventing recycling while some thread might be in the midst of one of the problem operations.  The only current user of this class is G1DirtyCardQueueSet, where a combination of GlobalCounter critical sections and safepoint boundaries are used to ensure ABA can't happen.

Reply:
Understood, but it still, to me, reads oddly to claim "we're done" and then have an ABA disclaimer.